pkg/convert/tf.go

```diff
@@ -1141,14 +1141,77 @@ func convertTemplateWrapExpr(state *convertState,
 	return tokens
 }
 
+func convertTemplateJoinExpr(state *convertState, inBlock bool, scopes *scopes,
+	fullyQualifiedPath string, expr *hclsyntax.TemplateJoinExpr,
+) hclwrite.Tokens {
+	forExpr := expr.Tuple.(*hclsyntax.ForExpr)
+	if forExpr == nil {
+		contract.Failf("Template Join Expression must contain for expression, but contains %T", expr.Tuple)
+	}
+
+	// The collection doesn't yet have access to the key/value scopes
+	collTokens := convertExpression(state, false, scopes, "", forExpr.CollExpr)
+
+	// TODO: We should ensure key and value vars are unique
+	locals := map[string]string{
+		forExpr.ValVar: camelCaseName(forExpr.ValVar),
+	}
+	if forExpr.KeyVar != "" {
+		locals[forExpr.KeyVar] = camelCaseName(forExpr.KeyVar)
+	}
+	scopes.push(locals)
+
+	valueTokens := convertExpression(state, false, scopes, "", forExpr.ValExpr)
+	valueTokens = valueTokens[1 : len(valueTokens)-1] // strip the outer quote tokens
+
+	scopes.pop()
+
+	var tokens hclwrite.Tokens
+	tokens = append(tokens, makeToken(hclsyntax.TokenTemplateControl, "%{"))
+
+	// Write the for template control section.
+	tokens = append(tokens, makeToken(hclsyntax.TokenIdent, "for"))
+	if locals[forExpr.KeyVar] != "" {
+		tokens = append(tokens, makeToken(hclsyntax.TokenIdent, locals[forExpr.KeyVar]))
+		tokens = append(tokens, makeToken(hclsyntax.TokenComma, ","))
+	}
+	tokens = append(tokens, makeToken(hclsyntax.TokenIdent, locals[forExpr.ValVar]))
+	tokens = append(tokens, makeToken(hclsyntax.TokenIdent, "in"))
+	tokens = append(tokens, collTokens...)
+
+	tokens = append(tokens, makeToken(hclsyntax.TokenTemplateControl, "~}"))
+
+	// There is no key expr in template join expressions.
+
+	// Write the value part
+	tokens = append(tokens, valueTokens...)
+
+	// There is no ellipsis in template join expressions.
+
+	// There is no conditional part in template join expressions.
+
+	// Write the endfor control section.
```

Comment:
These comments are (at least for me) generating more questions than they are answers. I wonder if it's worth instead adding a chunky-ish doc comment to the function that talks about what comes in, and what comes out, potentially with a sort of "code to code" example? So TF in that corresponds to a TemplateJoinExpr and tokens out that correspond to some PCL?

Reply:
It actually leaves it the same, I'll just add the syntax up there that's probably going to help